### ISSUE TITLE
Flag duplicate cases in match-case

### ISSUE THREAD

**Issue by excubo-jg** [microsoft/pyright#5921]
Pyright does not flag the duplicate cases in the code below. As cases do not fall-through in python, the code in the second instance will never be executed and should be flagged. The same holds true for if/elif-chains.

```
# pyright: reportUnnecessaryComparison=true

def test(i: int):
    match i:
        case 0:
            pass
        case 0:
            pass
        case 2:
            pass
        case _:
            pass

    if i == 0:
        pass
    elif i == 0:
        pass
    elif i == 2:
        pass
    else:
        pass
```

**Comment by erictraut** [comment]
Pyright has an option for flagging unnecessary comparisons based on type narrowing. You can enable this using `reportUnnecessaryComparison`.

This option doesn't look for specific values because, well, it's a type checker and evaluates types. It doesn't track values. In the case of `match` statements, those aren't even values; they're patterns. 

The feature you're looking for sounds more like something in the realm of a linter, not a type checker.

**Comment by excubo-jg** [comment]
Tracking values to do an in depth-assessment to me is a different use case and I would not expect a type checker to provide such functionality.

From a user perspective this feels a bit like a grey area though. Pyright in some instances checks patterns or evaluates values. If I am adding `i = 1` in the beginning of the code above, e.g. 

```
def test(i: int):
    i = 1
    match i:

```
then error messages are triggered which refer to a pattern mismatch or an impossible evaluation:
![image](https://github.com/microsoft/pyright/assets/52322342/2926e28c-99e5-4f7b-8df8-331a7be321ed)

If Pyright were to extend its checking to include a simple test for duplicate patterns that would catch errors in match-case structures which more often than not are simple comparisons to a int or a str representing the cases of e.g. a business logic. If these become too numerous, errors tend to sneak in when doing updates. From my experience a straightforward test would be a very welcome addition to the functionality.

Such a check would also identify unreachable code (in the duplicate case instance) and Pyright warns of stale code in other cases, too.
